Use smart pointers as necessary when creating RemoteVideoFrameProxy

squelart opened the pull request:
#### 4ee9cbf711936dacba4c1f80e020f32c5e0933ce
<pre>
Use smart pointers as necessary when creating RemoteVideoFrameProxy
<a href="https://bugs.webkit.org/show_bug.cgi?id=261659">https://bugs.webkit.org/show_bug.cgi?id=261659</a>
rdar://115634461

Reviewed by Ryosuke Niwa and Dan Glastonbury.

Added `protectedConnection()` in some classes for consistency and simpler changes.

* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.h:
(WebKit::GPUConnectionToWebProcess::protectedConnection):
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerManagerProxy.cpp:
(WebKit::RemoteMediaPlayerManagerProxy::createMediaPlayer):
* Source/WebKit/WebProcess/GPU/GPUProcessConnection.h:
(WebKit::GPUProcessConnection::protectedConnection):
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.cpp:
(WebKit::RemoteGraphicsContextGLProxy::paintCompositedResultsToVideoFrame):
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.cpp:
(WebKit::MediaPlayerPrivateRemote::videoFrameForCurrentTime):
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.h:
* Source/WebKit/WebProcess/GPU/media/cocoa/MediaPlayerPrivateRemoteCocoa.mm:
(WebKit::MediaPlayerPrivateRemote::pushVideoFrameMetadata):
* Source/WebKit/WebProcess/GPU/webrtc/LibWebRTCCodecs.cpp:
(WebKit::LibWebRTCCodecs::completedDecoding):

Canonical link: <a href="https://commits.webkit.org/268630@main">https://commits.webkit.org/268630@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8a0a32234ae2407ca59eb0f3a37f8b04738f77e6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/20194 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/20628 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/21256 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/22093 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/18844 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/23873 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/20792 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/20298 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/20413 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/20318 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/17550 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/22944 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/17487 "Passed tests") | | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/24619 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/18564 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/18526 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/22590 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/19113 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/16230 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/18318 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4854 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/22660 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/18938 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebKit/WebProcess/GPU/media/RemoteVideoFrameProxy.cpp:
```diff
@@ -55,28 +55,28 @@ RemoteVideoFrameProxy::Properties RemoteVideoFrameProxy::properties(WebKit::Remo
 }
 
-Ref<RemoteVideoFrameProxy> RemoteVideoFrameProxy::create(IPC::Connection& connection, RemoteVideoFrameObjectHeapProxy& videoFrameObjectHeapProxy, Properties&& properties)
+Ref<RemoteVideoFrameProxy> RemoteVideoFrameProxy::create(Ref<IPC::Connection> connection, Ref<RemoteVideoFrameObjectHeapProxy> videoFrameObjectHeapProxy, Properties&& properties)
```
These two changes seem unnecessary. Generally, we don't need to change the function arguments from raw reference/pointer to Ref/RefPtr.

squelart replied:
My idea was to make sure the caller would do the right thing implicitly, and in case of `create` functions and constructors there's no extra ref-count churn because the `Ref` gets moved into the object.
But as discussed, you've explained that the overall strategy is that callers are responsible for keeping function arguments alive (it's more consistent everywhere), so I'll redo this patch accordingly...

rniwa commented on Source/WebKit/WebProcess/GPU/media/RemoteVideoFrameProxy.cpp:
```diff
@@ -55,28 +55,28 @@ RemoteVideoFrameProxy::Properties RemoteVideoFrameProxy::properties(WebKit::Remo
 }
 
-static void releaseRemoteVideoFrameProxy(IPC::Connection& connection, const RemoteVideoFrameWriteReference& reference)
+static void releaseRemoteVideoFrameProxy(const Ref<IPC::Connection>& connection, const RemoteVideoFrameWriteReference& reference)
```
Same thing here. We don't need to change the argument type.

rniwa commented on Source/WebKit/WebProcess/GPU/media/RemoteVideoFrameProxy.cpp:
```diff
@@ -55,28 +55,28 @@ RemoteVideoFrameProxy::Properties RemoteVideoFrameProxy::properties(WebKit::Remo
 }
 
-void RemoteVideoFrameProxy::releaseUnused(IPC::Connection& connection, Properties&& properties)
+void RemoteVideoFrameProxy::releaseUnused(const Ref<IPC::Connection>& connection, Properties&& properties)
```
Ditto.

rniwa commented on Source/WebKit/WebProcess/GPU/media/RemoteVideoFrameProxy.cpp:
```diff
@@ -55,28 +55,28 @@ RemoteVideoFrameProxy::Properties RemoteVideoFrameProxy::properties(WebKit::Remo
 }
 
-RemoteVideoFrameProxy::RemoteVideoFrameProxy(IPC::Connection& connection, RemoteVideoFrameObjectHeapProxy& videoFrameObjectHeapProxy, Properties&& properties)
+RemoteVideoFrameProxy::RemoteVideoFrameProxy(Ref<IPC::Connection> connection, Ref<RemoteVideoFrameObjectHeapProxy> videoFrameObjectHeapProxy, Properties&& properties)
```
Ditto.

rniwa commented on Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.cpp:
```diff
@@ -257,7 +257,7 @@ RefPtr<WebCore::VideoFrame> RemoteGraphicsContextGLProxy::paintCompositedResults
     if (!result)
         return nullptr;
-    return RemoteVideoFrameProxy::create(WebProcess::singleton().ensureGPUProcessConnection().connection(), WebProcess::singleton().ensureGPUProcessConnection().videoFrameObjectHeapProxy(), WTFMove(*result));
+    return RemoteVideoFrameProxy::create(Ref { WebProcess::singleton().ensureGPUProcessConnection().connection() }, WebProcess::singleton().ensureGPUProcessConnection().videoFrameObjectHeapProxy(), WTFMove(*result));
```
Can we add protectedConnection to GPUProcessConnection instead?

djg reviewed (approved):
:shipit:

squelart committed:
Use smart pointers as necessary when creating RemoteVideoFrameProxy
https://bugs.webkit.org/show_bug.cgi?id=261659
rdar://115634461

Reviewed by Ryosuke Niwa and Dan Glastonbury.

Added `protectedConnection()` in some classes for consistency and simpler changes.

* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.h:
(WebKit::GPUConnectionToWebProcess::protectedConnection):
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerManagerProxy.cpp:
(WebKit::RemoteMediaPlayerManagerProxy::createMediaPlayer):
* Source/WebKit/WebProcess/GPU/GPUProcessConnection.h:
(WebKit::GPUProcessConnection::protectedConnection):
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.cpp:
(WebKit::RemoteGraphicsContextGLProxy::paintCompositedResultsToVideoFrame):
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.cpp:
(WebKit::MediaPlayerPrivateRemote::videoFrameForCurrentTime):
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.h:
* Source/WebKit/WebProcess/GPU/media/cocoa/MediaPlayerPrivateRemoteCocoa.mm:
(WebKit::MediaPlayerPrivateRemote::pushVideoFrameMetadata):
* Source/WebKit/WebProcess/GPU/webrtc/LibWebRTCCodecs.cpp:
(WebKit::LibWebRTCCodecs::completedDecoding):

Canonical link: https://commits.webkit.org/268630@main
